### PR TITLE
[storage/mmr] binary search in to_nearest_size

### LINKED ISSUE #820: [storage/mmr] log2(n) algo for finding nearest valid MMR size
In places where we have an integer N that we need to "round down" to the nearest valid MMR size, we currently iteratively call check_validity, and subtract 1 if it's false. This is a log2(n)^2 algo in the worst case. We should implement a function that yields the correct value directly in log2(n).

### PR DESCRIPTION
**Pull request by syrupipy**
Binary search on leaf count instead of iterating backwards. Drops complexity from O(log²n) to O(log n).

Fixes #820

### CODE REVIEW

File: storage/src/mmr/iterator.rs [hunk 1]
```diff
@@ -61,25 +61,49 @@ impl PeakIterator {
...
+                left = mid;
+            } else {
+                right = mid - 1;
+            }
         }
 
-        size
+        // left is the largest N where mmr_size(N) <= size
+        let result = 2 * left - left.count_ones() as u64;
+        Position::new(result)
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs]:
```suggestion
        Position::try_from(Location::new_unchecked(left).expect("shouldn't overflow")
```

File: storage/src/mmr/iterator.rs [hunk 2]
```diff
@@ -264,4 +288,87 @@ mod tests {
         let largest_valid_size = Position::new(u64::MAX >> 1);
         PeakIterator::to_nearest_size(largest_valid_size + Position::new(1));
     }
+
+    #[test]
+    fn test_to_nearest_size() {
+        // Test edge cases
+        assert_eq!(PeakIterator::to_nearest_size(Position::new(0)), 0);
+        assert_eq!(PeakIterator::to_nearest_size(Position::new(1)), 1);
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs]: Another edge case to test is the maximum allowed input (all bits other than topmost are 1):
`let largest_valid_size = Position::new(u64::MAX >> 1);`

File: storage/src/mmr/iterator.rs [hunk 3]
```diff
@@ -61,25 +61,49 @@ impl PeakIterator {
...
-            // A size-0 MMR is always valid so this loop must terminate before underflow.
-            size -= 1;
+        if size == 0 {
+            return size;
+        }
+
+        // Binary search for the largest N (number of leaves) such that
+        // mmr_size(N) = 2*N - popcount(N) <= size
+        let size_val = size.as_u64();
+        let mut left = 0u64;
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs]: I think something like lo/hi or low/high would be more descriptive than left/right.

File: storage/src/mmr/iterator.rs [hunk 4]
```diff
@@ -264,4 +288,87 @@ mod tests {
...
+        // Test some specific known cases
+        // MMR sizes by number of leaves (using formula: 2*N - popcount(N)):
+        // 0 leaves: size = 0
+        // 1 leaf:   size = 1
+        // 2 leaves: size = 3
+        // 3 leaves: size = 4
+        // 4 leaves: size = 7
+        // 5 leaves: size = 8
+
+        assert_eq!(PeakIterator::to_nearest_size(Position::new(0)), 0);
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs:343]: You could collapse this into a loop over `size` where you call `size.is_mmr_size()` to determine when the expected value should be bumped to match.

File: storage/src/mmr/iterator.rs (hunk 1 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: `Location::MAX` is 2^62-1 but leaf count can be up to 2^63-1, so this would overflow. Keeping the direct calculation to handle the full range. 
Am I right?

File: storage/src/mmr/iterator.rs (hunk 3 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: yeah much better

File: storage/src/mmr/iterator.rs (hunk 1 above)
Comment (roberto-bayardo) [reply storage/src/mmr/iterator.rs]: If there are more leaves than MAX_LOCATION then everything breaks.

Perhaps change the assertion condition on input size at the top to assert it's <= mmr::position::MAX_POSITION, which I think should guarantee we'll never produce a location > MAX_LOCATION? 

cc: @danlaine

File: storage/src/mmr/iterator.rs [hunk 5]
```diff
@@ -264,4 +290,82 @@ mod tests {
...
+        }
+
+        // Test with large value
+        let large_size = Position::new(1_000_000);
+        let rounded = PeakIterator::to_nearest_size(large_size);
+        assert!(rounded.is_mmr_size());
+        assert!(rounded <= large_size);
+
+        // Test maximum allowed input (all bits except topmost are 1)
+        let largest_valid_size = Position::new(u64::MAX >> 1);
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs]: Let's replace the u64::MAX >> 1 here with crate::mmr::MAX_POSITION

File: storage/src/mmr/iterator.rs [hunk 6]
```diff
@@ -61,25 +61,51 @@ impl PeakIterator {
...
+    /// mmr_size(N) <= size.
+    ///
     /// # Panics
     ///
     /// Panics if `size` is too large (specifically, the topmost bit should be 0).
-    pub fn to_nearest_size(mut size: Position) -> Position {
-        // If we don't check for overflow here, we could get stuck looping for a _very_ long time
-        // below, since we no longer have a log(n) bound on the number of iterations.
+    pub fn to_nearest_size(size: Position) -> Position {
         assert_ne!(size.leading_zeros(), 0, "overflow");
```
Comment (roberto-bayardo) [inline storage/src/mmr/iterator.rs]: While you're here let's update this to check that size <= crate::mmr::MAX_POSITION

File: storage/src/mmr/iterator.rs (hunk 5 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: Done! Updated to use `crate::mmr::MAX_POSITION`

File: storage/src/mmr/iterator.rs (hunk 1 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: Added the MAX_POSITION check as you suggested, ensures within safe bounds. should be better now

File: storage/src/mmr/iterator.rs [hunk 7]
```diff
@@ -61,25 +61,51 @@ impl PeakIterator {
...
+        // Binary search for the largest N (number of leaves) such that
+        // mmr_size(N) = 2*N - popcount(N) <= size
+        let size_val = size.as_u64();
+        let mut low = 0u64;
+        let mut high = size_val; // MMR size >= leaf count, so N <= size
+
+        while low < high {
+            // Use (low + high + 1) / 2 for upper-biased midpoint in binary search
+            #[allow(clippy::manual_div_ceil)]
+            let mid = (low + high + 1) / 2;
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]: I think we should use `div_ceil` here. I don't think that copilot comment was right. If `low == 0` and `high == 1` then `mid == 1` and we assign `low = mid == 1` and on the next iteration `low == high == 1` and we break and return `1`.

File: storage/src/mmr/iterator.rs [hunk 8]
```diff
@@ -61,25 +61,51 @@ impl PeakIterator {
...
+    ///
+    /// # Algorithm
+    ///
+    /// A valid MMR size corresponds to a specific number of leaves N, where:
+    /// `mmr_size(N) = 2*N - popcount(N)`
+    ///
+    /// This formula comes from the fact that N leaves require N-1 internal nodes, but merging
+    /// creates popcount(N)-1 additional nodes. We binary search for the largest N where
+    /// mmr_size(N) <= size.
+    ///
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]: Nit: Can we move this to be a comment underneath `pub fn to_nearest_size(size: Position) -> Position`?

This is an implementation detail that I don't think we need to expose in the doc comment.

File: storage/src/mmr/iterator.rs (hunk 7 above)
Comment (roberto-bayardo) [reply storage/src/mmr/iterator.rs]: chat gpt suggests:

`let mid = low + (high - low + 1) / 2;`

for the "upper biased, overflow proof solution"

File: storage/src/mmr/iterator.rs [hunk 9]
```diff
@@ -259,9 +285,86 @@ mod tests {
...
+                    rounded,
+                    test_pos,
+                    current_size
+                );
+
+                // Verify rounded is the largest valid size <= test_pos
+                if rounded < test_pos {
+                    assert!(
+                        !Position::new(rounded.as_u64() + 1).is_mmr_size()
+                            || rounded.as_u64() + 1 > test_pos,
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]: if `rounded < test_pos` then `rounded + 1 > test_pos` never holds

File: storage/src/mmr/iterator.rs [hunk 10]
```diff
@@ -259,9 +285,86 @@ mod tests {
...
+                    "rounded {} should be <= test_pos {} (current: {})",
+                    rounded,
+                    test_pos,
+                    current_size
+                );
+
+                // Verify rounded is the largest valid size <= test_pos
+                if rounded < test_pos {
+                    assert!(
+                        !Position::new(rounded.as_u64() + 1).is_mmr_size()
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]:
```suggestion
                        !(rounded + 1).is_mmr_size()
```

File: storage/src/mmr/iterator.rs [hunk 11]
```diff
@@ -259,9 +285,86 @@ mod tests {
...
+        // Build an MMR incrementally and verify to_nearest_size for all intermediate values
+        let mut mmr: Mmr<Sha256> = Mmr::new();
+        let mut hasher = Standard::<Sha256>::new();
+        let digest = [1u8; 32];
+
+        for _ in 0..1000 {
+            let current_size = mmr.size();
+
+            // Test all positions from previous size to current size
+            for test_pos in *current_size..=*current_size + 10 {
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]: This comment doesn't match the iterations bounds

File: storage/src/mmr/iterator.rs [hunk 12]
```diff
@@ -259,9 +285,86 @@ mod tests {
...
-        let largest_valid_size = Position::new(u64::MAX >> 1);
-        PeakIterator::to_nearest_size(largest_valid_size + Position::new(1));
+        PeakIterator::to_nearest_size(crate::mmr::MAX_POSITION + Position::new(1));
+    }
+
+    #[test]
+    fn test_to_nearest_size() {
+        // Test edge cases
+        assert_eq!(PeakIterator::to_nearest_size(Position::new(0)), 0);
+        assert_eq!(PeakIterator::to_nearest_size(Position::new(1)), 1);
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]: This should move to `test_to_nearest_size_specific_cases`

File: storage/src/mmr/iterator.rs [hunk 13]
```diff
@@ -259,9 +285,86 @@ mod tests {
     }
 
     #[test]
-    #[should_panic]
+    #[should_panic(expected = "size exceeds MAX_POSITION")]
     fn test_to_nearest_size_panic() {
-        let largest_valid_size = Position::new(u64::MAX >> 1);
-        PeakIterator::to_nearest_size(largest_valid_size + Position::new(1));
+        PeakIterator::to_nearest_size(crate::mmr::MAX_POSITION + Position::new(1));
```
Comment (danlaine) [inline storage/src/mmr/iterator.rs]:
```suggestion
        PeakIterator::to_nearest_size(crate::mmr::MAX_POSITION +1);
```

File: storage/src/mmr/iterator.rs (hunk 7 above)
Comment (danlaine) [reply storage/src/mmr/iterator.rs]: I think that will also return the same thing as `(low + high).div_ceil(2)`?

Comment (roberto-bayardo) [reply storage/src/mmr/iterator.rs]: same as `low + (high - low).div_ceil(2)` I guess.

File: storage/src/mmr/iterator.rs (hunk 8 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: Sure

File: storage/src/mmr/iterator.rs (hunk 11 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: nice catch

File: storage/src/mmr/iterator.rs (hunk 7 above)
Comment (syrupipy) [reply storage/src/mmr/iterator.rs]: Anyway let me revert `div_ceil ` at present

**Review by roberto-bayardo — APPROVED** [review]
thanks!

**Comment by roberto-bayardo** [comment]
looks like you've got one lint error then should be good to go:
```
cargo +nightly fmt --all -- --check
Diff in /home/runner/work/monorepo/monorepo/storage/src/mmr/iterator.rs:69:
     ///
     /// Panics if `size` exceeds [crate::mmr::MAX_POSITION].
     pub fn to_nearest_size(size: Position) -> Position {
-        assert!(size <= crate::mmr::MAX_POSITION, "size exceeds MAX_POSITION");
+        assert!(
+            size <= crate::mmr::MAX_POSITION,
+            "size exceeds MAX_POSITION"
+        );
 
         // Algorithm: A valid MMR size corresponds to a specific number of leaves N, where:
         // mmr_size(N) = 2*N - popcount(N)
error: Recipe `check-fmt` failed on line 26 with exit code 1
Error: Process completed with exit code 1.
```
